File: NeetCode/1-Arrays & Hashing/4-Group Anagrams/49-Group-Anagrams.ts
// function sortString(str: string): string {
//   return str.split("").sort().join("");
// }

// function groupAnagrams(strs: string[]): string[][] {
//   const store = {};
//   const result: string[][] = [];
//   for (let i = 0; i < strs.length; i++) {
//     const sortedStr = sortString(strs[i]);
//     if (store[sortedStr]) store[sortedStr].push(strs[i]);
//     else store[sortedStr] = [strs[i]];
//   }
//   for (let key in store) result.push(store[key]);
//   return result;
// }

function makeKey(str:string):string{
    const count=new Array(26).fill(0);
    for(let c of str)count[c.charCodeAt(0)-'a'.charCodeAt(0)]+=1;
    return count.join(',');
}

function groupAnagrams(strs: string[]): string[][] {
  const store={};
  for(let str of strs ){
    const key=makeKey(str);
    if(!store[key])store[key]=[];
    store[key].push(str);
  }
  return Object.values(store);
}

File: NeetCode/1-Arrays & Hashing/4-Group Anagrams/49-Group-Anagrams.cpp
class Solution {
public:
    string makeKey(string str){
        vector<int>count(26,0);
        for(auto c : str){
            count[c-'a']++;
        }
        string key=to_string(count[0]);
        for (int i = 1; i < 26; i++) {
                key += ',' + to_string(count[i]);
            }
            return key;
    }
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string,vector<string>>store;
        for(auto str : strs){
            string key=makeKey(str);
            store[key].push_back(str);
        }
        vector<vector<string>>result;
        for(auto it=store.begin();it!=store.end();it++){
            result.push_back(it->second);
        }
        return result;
    }
};